docs/proposals/devfile.md

```diff
@@ -0,0 +1,64 @@
+
+# Devfile support in odo
+
+To better integrate odo with other developer tooling we need to align on one common definition of a developer workspace and application lifecycle.
+
+[Eclipse Che](https://github.com/eclipse/che) project already created and started using [Devfile](https://github.com/redhat-developer/devfile). Devfile is a YAML file that defines developer workspace in Che environment. It is structured in such a way that can be used also in odo.
+
+Implementing Devfile support in odo will create a great benefit for users, as it will allow them to easily switch tools with no additional configuration. Adding support for new languages will be also much easier with devfiles, as it will be only a matter of using the right devfile template.
+
+## Implementation plan
+
+### Consuming current version of Devfile in odo
+
+1. Create a golang library for parsing and reading data from Devfile
+    * Devfile is versioned file so the design of a library needs to be able to handle and support multiple Devfile versions at the same time
+
+        **Parser high level design**
+        * The parser would read the provided devfile.
+        * From the devfile, parser would look for the API Version of the devfile and validate whether the API version in supported in odo.
+        * The devfile schema is based on <http://json-schema.org/draft-07/schema#>
+        * Hence, the parser would validate the integrity of the provided devfile with the devfile json schema of the respective API version.
+        * Furthermore, parser would try to map devfile sections (projects, components, commands etc) to their respective Golang structs.
+        * The idea is to have same Golang structs for the devfile sections across different API version contingent to the condition that the sections themselves are compatible across the API versions..
+        * This is very similar to the way Golang structs are handled across different GVK versions in kubernetes. [https://github.com/kubernetes/kubernetes/blob/ded2ff39c381b122951dd103dafb291755791921/staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go#L32](https://github.com/kubernetes/kubernetes/blob/ded2ff39c381b122951dd103dafb291755791921/staging/src/k8s.io/apimachinery/pkg/runtime/scheme.go#L32)
+        * The implementation details for this with respect to odo are being worked upon and once confirmed would be updated here.
+
+2. Implement experimental support for reading current version of a Devfile and bootstrap odo components from a Devfile.
+    * If there is a devfile.yml and no .odo/config.yaml `odo push` will use Devfile.yaml and create necessary resources in Kubernetes cluster.
+    * Before devfile has support for category of a commands (application lifecycle commands like run, build, test etc..) we could just add flags to odo push that will allow to specify the name of the command that should be used for build and run, as currently those are only two commands that odo cares about. \
```

Comment:
If it's not too much of a hassle, we could try adding support for category commands instead of adding flags to `odo push` because making changes later will also mean that plugins dependant on odo would have to change the way things are done. This can cause more breakage in those tools than we can initially estimate.

Reply:
Any changes to devfile need to be consulted with Che team and it Che needs to support those also.
Devfile implementation will have the status of an experiment (probably hidden behind feature flag) until we have a stable interface.

Reply:
I guess there's some misunderstanding here. I understand that changes to DevFile need to be discussed with Che team. But I'm not proposing a change to DevFile here.

I'm just asking to add support for category commands to odo right away instead of having flags for it on a temporary basis. So, odo should be able to understand category commands in the DevFile (if they are already a part of current version of DevFile we're developing against.)

Reply:
That is the problem. There is no such thing as category of commands in the current devfile.
The goal is to make odo work with the current devfiles version.
I don't see any other option than adding the flags until devfile has proper support for this.

We could try to do some smart autodetection based on the command name. Like if the command name exactly matches "build" we assume that is built command and the same for "run".  But there is nothing in devfile spec that dictates how the name should look like, it is just free form string. That was originally meant for displaying in the UI for the user to select one.
Even if we try to do this smart detection we will need the flags to allow user overwrite it if is wrong.

We can discuss this in later stages.